File: src/_experiments/verlet.rb
require 'jruby_art'
require 'jruby_art/app'
require 'toxiclibs'

require_relative '../_lib/pflib/pflib'

class Verlet < PF::ProcessingApp
    attr_accessor :circle

    #include Physics
    load_library :controlP5
    include_package 'controlP5'
    include_package 'toxi.geom'
    include_package 'toxi.physics2d'
    include_package 'toxi.physics2d.constraints'

    def settings
        size 500, 500
    end

    def setup
        super

        no_loop

        @circle = Toxi::Circle.new width/2, height/2, 300

        @ph = VerletPhysics2D.new
        @ph.set_index(PointQuadtree.new(0, 0, width, height))

        # @ph.add_constraint(Class.new do
        #    include ParticleConstraint2D
        #    def apply(p)
        #        unless $app.circle.contains_point p
        #            p.set($app.circle.add(p.sub($app.circle).normalize_to($app.circle.get_radius)));
        #        end
        #    end
        # end.new)
        #@ph.add_constraint(Class.new do
        #    include ParticleConstraint2D
        #    def apply(p)
        #        if $app.circle.contains_point p
        #            tan = $app.circle.get_tangent_points(p)
        #            p.set(p.distance_to(tan[0]) < p.distance_to(tan[1]) ? tan[0] : tan[1]) if tan
        #        end
        #    end
        #end.new)
    end

    def draw
        clear

        @ph.update


        fill 0, 255, 0, 5
        stroke_width 2
        stroke 0, 255, 0, 20
        ellipse @circle.x, @circle.y, @circle.get_radius, @circle.get_radius

        within = @ph.get_index.items_within_radius(TVec2D.new(@circle.x, @circle.y), @circle.get_radius/2, [])
        @ph.particles.each do |p|
            if within.include? p
                fill 0, 255, 0
                stroke_width 4
                stroke 0, 255, 0, 20
            else
                assert (@ph.particles - within).include? p
                fill 255, 0, 0
                stroke_width 4
                stroke 0, 20
            end
            ellipse p.x, p.y, 6, 6
        end

        sketch_title "#{frame_rate.round.to_s}fps"
        add_particles
        loop
    end

    def add_particles
        1.times do
            p = VerletParticle2D.new(rand(0..width), rand(0..height))
            @ph.add_particle(p)
            #@ph.particles.each do |pp|
            #    @ph.add_spring(LineSpring.new(p, pp))
            #end
        end
        redraw
    end

    def mouse_pressed
        add_particles
    end

    def key_pressed(e)
        case key
            when '1', '2'
                add_particles
            when 'q', 'w'
                @circle.set_radius(@circle.get_radius + 1)
            when 'a', 's'
            when 'z', 'x'
            else
                super
        end

        sketch_title "#{paused? ? 'paused' : ''}"
        redraw
    end

    def build_gui
        super do |cp|
            #cp.add_slider('a')
            #    .set_value(0.0)
            #    .set_range(0.0, 2*PI)
            #    .set_number_of_tick_marks(40)
            #    .snap_to_tick_marks(true)
            #    .set_height(19)
            #    .update_size
            #    .on_change { |e| @phi = e.get_controller.get_value }
        end

    end
end

Verlet.new unless defined? $app

File: src/java/_experiments/Verlet.java
package _experiments;

import processing.core.PApplet;
import toxi.geom.Circle;
import toxi.geom.PointQuadtree;
import toxi.geom.Rect;
import toxi.geom.SpatialIndex;
import toxi.geom.Vec2D;
import toxi.physics2d.VerletParticle2D;
import toxi.physics2d.VerletPhysics2D;
import toxi.physics2d.behaviors.ParticleBehavior2D;
import toxi.processing.ToxiclibsSupport;

public class Verlet extends PApplet {
	private VerletPhysics2D ph = new VerletPhysics2D();
	private Circle circle;
	private ToxiclibsSupport gfx;
	private static final int EFFECT_RADIUS = 80;

	@Override
	public void settings() {
		size(500, 500);
	}

	@Override
	public void setup() {
		super.setup();

		background(255);
		gfx = new ToxiclibsSupport(this);

		ph.setWorldBounds(new Rect(0, 0, width, height));
		ph.setIndex(new PointQuadtree(0, 0, width, height));


		addParticles();
		//noLoop();
	}

	private void addParticles() {
		Circle circle = new Circle(width/2, height/2, 50);
		circle.toPolygon2D(10).vertices.stream()
				.map(VerletParticle2D::new)
				.forEach(p -> {
					p.addBehavior(new ConstantAttractionBehavior2D(p, 80, 0.5f));
					p.addBehavior(new ConstantAttractionBehavior2D(p, 120, -0.6f));
					ph.addParticle(p);
				});
		ph.particles.get(6).x += 0.2;
	}

	@Override
	public void keyPressed() {
		//addParticles();
		redraw();
	}

	@Override
	public void draw() {
		background(255);

		/*
		strokeWeight(1);
		stroke(0, 20);
		fill(0, 5);
		gfx.ellipse(circle);
		*/

		strokeWeight(4);
		stroke(0, 100);
		fill(0, 240);
		ph.particles.forEach(p -> {
			ellipse(p.x, p.y, 4, 4);
		});

		strokeWeight(1);
		stroke(255, 0, 0, 20);
		noFill();

		for (int i=0; i < ph.particles.size(); i++) {
			VerletParticle2D a = ph.particles.get(i);
			if (i == 6) {
				ellipse(a.x, a.y, 80, 80);
				ellipse(a.x, a.y, 120, 120);
			}
			for (int j=i+1; j < ph.particles.size(); j++) {
				VerletParticle2D b = ph.particles.get(j);
				line(a.x, a.y, b.x, b.y);
			}
		}

		ph.update();

	}

	static class ConstantAttractionBehavior2D implements ParticleBehavior2D {
		private VerletParticle2D a;
		private float distance = 0;
		private float radius = 0;

		public ConstantAttractionBehavior2D(VerletParticle2D a, float radius, float distance) {
			this.a = a;
			this.distance = distance;
			this.radius = radius;
		}

		@Override
		public void apply(VerletParticle2D p) {
			a.set(a.interpolateTo(p, distance));
		}

		@Override
		public void applyWithIndex(SpatialIndex<Vec2D> index) {
			index.itemsWithinRadius(a, radius, null).stream()
					.map(v -> (VerletParticle2D)v)
					.forEach(this::apply);
		}

		@Override
		public void configure(float timeStep) {
		}

		@Override
		public boolean supportsSpatialIndex() {
			return true;
		}
	}

	public static void main(String[] args) {
		PApplet.main("_experiments.verlet.Verlet");
	}
}